File: packages/db/src/models/Post.ts
import { Schema, type InferSchemaType, type Types } from "mongoose";
import { getModel } from "../lib/model.js";

const PostSchema = new Schema(
  {
    authorId: { type: Schema.Types.ObjectId, ref: "User", required: true },
    content: { type: String, required: true, maxlength: 280 },
    mediaUrls: [{ type: String }],
    replyToId: { type: Schema.Types.ObjectId, ref: "Post", default: null },
    repostOfId: { type: Schema.Types.ObjectId, ref: "Post", default: null },
    orbitId: { type: Schema.Types.ObjectId, ref: "Orbit", default: null },
    hashtags: [{ type: String }],
    mentions: [{ type: Schema.Types.ObjectId, ref: "User" }],
    stats: {
      likeCount: { type: Number, default: 0 },
      replyCount: { type: Number, default: 0 },
      repostCount: { type: Number, default: 0 },
      bookmarkCount: { type: Number, default: 0 },
      viewCount: { type: Number, default: 0 },
    },
    isDeleted: { type: Boolean, default: false },
  },
  { timestamps: true },
);

PostSchema.index({ authorId: 1, createdAt: -1 });
PostSchema.index({ orbitId: 1, createdAt: -1 });
PostSchema.index({ replyToId: 1 });
PostSchema.index({ hashtags: 1 });

export type IPost = InferSchemaType<typeof PostSchema> & {
  _id: Types.ObjectId;
  createdAt: Date;
  updatedAt: Date;
};

export const Post = getModel<IPost>("Post", PostSchema);

File: apps/api/app/models/post.py
from datetime import datetime

from beanie import Document, PydanticObjectId
from pydantic import BaseModel, Field


class PostStats(BaseModel):
    likeCount: int = 0
    replyCount: int = 0
    repostCount: int = 0
    bookmarkCount: int = 0
    viewCount: int = 0


class Post(Document):
    authorId: PydanticObjectId
    content: str
    mediaUrls: list[str] = Field(default_factory=list)
    replyToId: PydanticObjectId | None = None
    repostOfId: PydanticObjectId | None = None
    orbitId: PydanticObjectId | None = None
    hashtags: list[str] = Field(default_factory=list)
    mentions: list[PydanticObjectId] = Field(default_factory=list)
    stats: PostStats = Field(default_factory=PostStats)
    isDeleted: bool = False
    createdAt: datetime | None = None
    updatedAt: datetime | None = None

    class Settings:
        name = "posts"
